cloudinit/stages.py

```diff
@@ -710,26 +765,36 @@ def _apply_netcfg_names(self, netcfg):
             LOG.warning("Failed to rename devices: %s", e)
 
     def apply_network_config(self, bring_up):
-        # get a network config
+        """Apply the network config.
+
+        Find the config, determine whether to apply it, apply it via
+        the distro, and optionally bring it up
+        """
         netcfg, src = self._find_networking_config()
         if netcfg is None:
             LOG.info("network config is disabled by %s", src)
             return
 
-        # request an update if needed/available
-        if self.datasource is not NULL_DATA_SOURCE:
-            if not self.is_new_instance():
-                if not self.datasource.update_metadata([EventType.BOOT]):
-                    LOG.debug(
-                        "No network config applied. Neither a new instance"
-                        " nor datasource network update on '%s' event",
-                        EventType.BOOT)
-                    # nothing new, but ensure proper names
-                    self._apply_netcfg_names(netcfg)
-                    return
-                else:
-                    # refresh netcfg after update
-                    netcfg, src = self._find_networking_config()
+        def _boot_event_enabled_and_supported():
+            return self.update_event_enabled(
+                EventType.BOOT, scope=EventScope.NETWORK
+            ) and self.datasource.update_metadata([EventType.BOOT])
```

Comment:
This call still throws me off: this isn't just checking for support, but will actually cause the datasource to discard cached data and fetch it anew. That feels to me like it should be two separate calls: "does this datasource support $event_type" and "update the datasource's metadata for $event_type".

It also feels a little odd that `update_metadata` doesn't take a scope: we only want to update network metadata here, if (e.g.) storage metadata has changed then we don't particularly want to update it.

Relatedly, it may look like `update_metadata` should just update network metadata (or, generally, metadata for the explicitly configured scopes), but our `_get_data` implementations generally fetch and store All The Metadata, so I think this will completely refresh all metadata regardless: perhaps not a problem, but worth noting.